aws/resource_aws_autoscaling_group.go

```diff
@@ -1369,6 +1447,148 @@ func resourceAwsAutoscalingGroupDelete(d *schema.ResourceData, meta interface{})
 	return nil
 }
 
+func resourceAutoScalingGroupWarmPoolDelete(g *autoscaling.Group, d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).autoscalingconn
+
+	if g.WarmPoolConfiguration == nil {
+		// No warm pool configured. Skipping deletion.
+		return nil
+	}
+
+	log.Printf("[INFO] Group has a warming pool. First deleting it.")
+
+	if err := resourceAwsAutoscalingGroupWarmPoolDrain(d, meta); err != nil {
+		return err
+	}
+
+	// Delete Warm Pool if it is not pending delete.
+	if g.WarmPoolConfiguration.Status == nil || aws.StringValue(g.WarmPoolConfiguration.Status) != "PendingDelete" {
+		deleteopts := autoscaling.DeleteWarmPoolInput{
+			AutoScalingGroupName: aws.String(d.Id()),
+			ForceDelete:          aws.Bool(d.Get("force_delete").(bool) || d.Get("force_delete_warm_pool").(bool)),
+		}
+
+		err := resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+			if _, err := conn.DeleteWarmPool(&deleteopts); err != nil {
+				if callerr, ok := err.(awserr.Error); ok {
+					switch callerr.Code() {
+					case "ResourceInUse", "ScalingActivityInProgress":
+						// These are retryable
+						return resource.RetryableError(callerr)
+					}
+				}
+				// Didn't recognize the error, so shouldn't retry.
+				return resource.NonRetryableError(err)
+			}
+			// Successful delete
+			return nil
+		})
+
+		if isResourceTimeoutError(err) {
+			_, err = conn.DeleteWarmPool(&deleteopts)
+		}
+		if err != nil {
+			return fmt.Errorf("Error deleting Warm Pool: %s", err)
+		}
+	}
+
+	// Wait for Warm pool to be gone.
+	err := resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
```

Comment:
This looks like it could be handled with waiter. It would achieve the same, but is more uniform with other resources. See https://github.com/hashicorp/terraform-provider-aws/blob/03072134f5110be18b431bdb2d3e1758793098f9/aws/resource_aws_rds_global_cluster.go#L445-L465 for an example.

Reply:
Its possible a waiter may not work in this case if the state can't be used to determine if the warm pool removal is complete.

Reply:
Updated this to use waiter.